src/iterator.rs

```diff
@@ -431,13 +417,154 @@ impl<'a> Iterator for Forever<'a> {
     }
 }
 
+#[cfg(not(windows))]
+mod waker_impl {
+    use libc;
+    use std::io::Error;
+    use std::os::unix::io::AsRawFd;
+    use std::os::unix::io::RawFd;
+    use std::os::unix::net::UnixStream;
+
+    #[cfg(feature = "mio-support")]
+    use mio::unix::EventedFd;
+
+    use pipe;
+
+    #[derive(Debug)]
+    pub(super) struct Waker {
+        read: UnixStream,
+        write: UnixStream,
+        read_fd: RawFd,
+    }
+
+    impl Waker {
+        pub(super) fn new() -> Result<Self, Error> {
+            let (read, write) = UnixStream::pair()?;
+            let read_fd = read.as_raw_fd();
+            Ok(Self {
+                read,
+                write,
+                read_fd,
+            })
+        }
+
+        /// Sends a wakeup signal to the internal wakeup pipe.
+        pub(super) fn wake(&self) {
+            pipe::wake(self.write.as_raw_fd());
+        }
+
+        /// Reads data from the internal self-pipe.
+        ///
+        /// If `wait` is `true` and there are no data in the self pipe, it blocks until some come.
+        ///
+        /// Returns weather it successfully read something.
+        pub(super) fn flush(&self, wait: bool) -> bool {
+            const SIZE: usize = 1024;
+            let mut buff = [0u8; SIZE];
+            let res = unsafe {
+                // We ignore all errors on purpose. This should not be something like closed file
+                // descriptor. It could EAGAIN, but that's OK in case we say MSG_DONTWAIT. If it's
+                // EINTR, then it's OK too, it'll only create a spurious wakeup.
+                libc::recv(
+                    self.read.as_raw_fd(),
+                    buff.as_mut_ptr() as *mut libc::c_void,
+                    SIZE,
+                    if wait { 0 } else { libc::MSG_DONTWAIT },
+                )
+            };
+
+            res > 0
+        }
+
+        #[cfg(feature = "mio-support")]
+        pub(super) fn evented(&self) -> EventedFd {
+            EventedFd(&self.read_fd)
+        }
+    }
+}
+
+#[cfg(windows)]
+mod waker_impl {
+    // Unlike `sigaction`, `SetConsoleCtrlHandler` launches a separate thread for handlers.
+    // Therefore we don't need to be AS-safe and can just use an ordinary in-process synchronization primitives.
+
+    use std::io::Error;
+    use std::mem;
+    use std::sync::{Condvar, Mutex};
```

Comment:
I thought `Mutex` with `Condvar` is "enough" because it's more primitive. But I see that `mpsc` might be more suitable here because it's parallel with `#[cfg(unix)]` implementaion.